src/java_bytecode/java_bytecode_convert_class.cpp

```diff
@@ -517,3 +517,150 @@ bool java_bytecode_convert_class(
+/// \param type
+/// \param replacement_parameters
+void find_and_replace_parameters(
+  typet &type,
+  const std::vector<java_generic_parametert> &replacement_parameters)
+{
+  if(is_java_generic_parameter(type))
+  {
+    find_and_replace_parameter(
+      to_java_generic_parameter(type), replacement_parameters);
+  }
+  else if(is_java_generic_type(type))
+  {
+    java_generic_typet &generic_type = to_java_generic_type(type);
+    std::vector<reference_typet> &arguments =
+      generic_type.generic_type_arguments();
+    for(auto argument : arguments)
+    {
+      find_and_replace_parameters(argument, replacement_parameters);
+    }
+  }
+}
+
+/// Checks if the class is implicitly generic, i.e., it is an inner class of
+/// any generic class. All uses of the implicit generic type parameters within
+/// the inner class are updated to point to the type parameters of the
+/// corresponding outer classes.
+/// \param class_name
+/// \param symbol_table
+void mark_java_implicitly_generic_class_type(
+  const irep_idt &class_name,
+  symbol_tablet &symbol_table)
+{
+  const std::string qualified_class_name = "java::" + id2string(class_name);
+  PRECONDITION(symbol_table.has_symbol(qualified_class_name));
+
+  // create a vector of all generic type parameters of all outer classes, in
+  // the order from the outer-most inwards
+  std::vector<java_generic_parametert> implicit_generic_type_parameters;
+  std::string::size_type outer_class_delimiter =
+    qualified_class_name.rfind("$");
+  while(outer_class_delimiter != std::string::npos)
+  {
+    std::string outer_class_name =
+      qualified_class_name.substr(0, outer_class_delimiter);
+    if(symbol_table.has_symbol(outer_class_name))
+    {
+      const symbolt &outer_class_symbol =
+        symbol_table.lookup_ref(outer_class_name);
+      const java_class_typet &outer_class_type =
+        to_java_class_type(outer_class_symbol.type);
+      if(is_java_generic_class_type(outer_class_type))
+      {
+        const auto &outer_generic_type_parameters =
+          to_java_generic_class_type(outer_class_type).generic_types();
+        implicit_generic_type_parameters.insert(
+          implicit_generic_type_parameters.begin(),
+          outer_generic_type_parameters.begin(),
+          outer_generic_type_parameters.end());
+      }
+      outer_class_delimiter = outer_class_name.rfind("$");
+    }
+    else
+    {
+      throw missing_outer_class_symbol_exceptiont(
+        outer_class_name, qualified_class_name);
+    }
+  }
+
+  // if there are any implicit generic type parameters, mark the class
+  // implicitly generic and update identifiers of type parameters used in fields
+  if(!implicit_generic_type_parameters.empty())
+  {
+    symbolt &class_symbol =
+      symbol_table.get_writeable_ref(qualified_class_name);
+    java_class_typet &class_type = to_java_class_type(class_symbol.type);
+
+    class_type = java_implicitly_generic_class_typet(
```

Comment:
Recommend assigning to `class_symbol.type`. The effect is the same, but it'll be clearer that you're overwriting the symbol type not a local.